geozero/src/gpx/gpx_reader.rs

```diff
@@ -0,0 +1,216 @@
+use crate::error::GeozeroError;
+use std::io;
+
+pub fn read_gpx<R: io::Read, P: crate::GeomProcessor>(
+    reader: &mut R,
+    processor: &mut P,
+) -> crate::error::Result<()> {
+    let gpx_reader = match gpx::read(reader) {
+        Ok(r) => r,
+        Err(e) => return Err(GeozeroError::Geometry(e.to_string())),
+    };
+
+    process_top_level_waypoints(&gpx_reader, processor)?;
+    process_top_level_tracks(&gpx_reader, processor)?;
+    process_top_level_routes(&gpx_reader, processor)?;
+
+    Ok(())
+}
+
+fn process_top_level_waypoints<P: crate::GeomProcessor>(
+    gpx_reader: &gpx::Gpx,
+    processor: &mut P,
+) -> crate::error::Result<()> {
+    if gpx_reader.waypoints.is_empty() {
+        return Ok(());
+    }
+    processor.multipoint_begin(gpx_reader.waypoints.len(), 0)?;
+    process_waypoints_iter(gpx_reader.waypoints.iter(), processor, 0)?;
```

Comment:
Should this `index` always be `0`?

Reply:
`index` is the nth item within a parent item:
```md
                          +-----------------+
                          |                 |
                          |     Initial     <-------------+
                          |                 |             |
                          +--------^--------+    +--------v---------+
                                   |             |                  |
                             +-+-+---+-+         |GeometryCollection|
                             | | | | | |         |                  |
                             v v v v v v         +--------^---------+
+-----------------+                                       |
|                 |                                 +-+-+---+-+
|  MultiPolygon   |                                 | | | | | |
|                 |                                 v v v v v v
+--------^--------+   +-----------------+
         |            |                 |
         |            | MultiLineString |
+--------v--------+   |                 |
|                 |   +--------^--------+
|     Polygon     |            |
|                 |            |
+--------^--------+   +--------v--------+                         +-------------+   +--------------+
         |            |                 |                         |             |   |              |
         +----------->+    LineString   |                         |    Point    |   |  MultiPoint  |
                      |                 |                         |             |   |              |
                      +--------+--------+   +-----------------+   +------+------+   +-------+------+
                               |            |                 |          |                  |
                               +------------>   Coordinate    <----------+------------------+
                                            |                 |
                                            +-----------------+
```
E.g. the nth `LineString` within a `MultiLineString` or the nth `Coordinate` within a `LineString`

Reply:
@pka Thanks for the explanation. Does the latest commit look correct? https://github.com/georust/geozero/pull/48/commits/80e82e33b019096e0cfb8b8ef62156b2886eaeeb